crates/red_knot_module_resolver/src/path.rs

```diff
@@ -1,676 +1,286 @@
-            ModulePathBufInner::FirstParty(path) => ModulePathRefInner::FirstParty(path),
-            ModulePathBufInner::StandardLibrary(FilePath::System(path)) => {
-                ModulePathRefInner::StandardLibrary(FilePathRef::System(path))
-            }
-            ModulePathBufInner::StandardLibrary(FilePath::Vendored(path)) => {
-                ModulePathRefInner::StandardLibrary(FilePathRef::Vendored(path))
-            }
-            ModulePathBufInner::SitePackages(path) => ModulePathRefInner::SitePackages(path),
-            ModulePathBufInner::EditableInstall(path) => ModulePathRefInner::EditableInstall(path),
-        };
-        ModulePathRef(inner)
-    }
-}
-
-impl PartialEq<SystemPath> for ModulePathRef<'_> {
-    fn eq(&self, other: &SystemPath) -> bool {
-        match self.0 {
-            ModulePathRefInner::Extra(path) => path == other,
-            ModulePathRefInner::FirstParty(path) => path == other,
-            ModulePathRefInner::SitePackages(path) => path == other,
-            ModulePathRefInner::EditableInstall(path) => path == other,
-            ModulePathRefInner::StandardLibrary(FilePathRef::System(path)) => path == other,
-            ModulePathRefInner::StandardLibrary(FilePathRef::Vendored(_)) => false,
+    pub(crate) fn with_py_extension(&self) -> Option<Self> {
+        if self.is_standard_library() {
+            return None;
         }
+        let ModulePath {
+            search_path,
+            relative_path,
+        } = self;
+        Some(ModulePath {
+            search_path: search_path.clone(),
+            relative_path: relative_path.with_extension("py"),
+        })
     }
 }
 
-impl PartialEq<ModulePathRef<'_>> for SystemPath {
-    fn eq(&self, other: &ModulePathRef) -> bool {
-        other == self
-    }
-}
-
-impl PartialEq<SystemPathBuf> for ModulePathRef<'_> {
+impl PartialEq<SystemPathBuf> for ModulePath {
     fn eq(&self, other: &SystemPathBuf) -> bool {
-        self == &**other
-    }
-}
-
-impl PartialEq<ModulePathRef<'_>> for SystemPathBuf {
-    fn eq(&self, other: &ModulePathRef<'_>) -> bool {
-        &**self == other
-    }
-}
-
-impl PartialEq<VendoredPath> for ModulePathRef<'_> {
-    fn eq(&self, other: &VendoredPath) -> bool {
-        match self.0 {
-            ModulePathRefInner::Extra(_) => false,
-            ModulePathRefInner::FirstParty(_) => false,
-            ModulePathRefInner::SitePackages(_) => false,
-            ModulePathRefInner::EditableInstall(_) => false,
-            ModulePathRefInner::StandardLibrary(FilePathRef::System(_)) => false,
-            ModulePathRefInner::StandardLibrary(FilePathRef::Vendored(path)) => path == other,
-        }
+        let ModulePath {
+            search_path,
+            relative_path,
+        } = self;
+        search_path
+            .as_system_path_buf()
+            .is_some_and(|search_path| search_path.join(relative_path) == *other)
```

Comment:
It's slightly surprising that `Eq` allocates. I think you could optimize this by doing 

```
if let Some(rest) = other.split_prefix(self.search_path) {
    rest == &self.relative_path
} else {
    false
}
```

Same for vendored path